lib/MySQL_Session.cpp

```diff
@@ -2984,113 +2467,6 @@ bool MySQL_Session::handler_again___status_CHANGING_USER_SERVER(int *_rc) {
 	return false;
 }
 
-bool MySQL_Session::handler_again___status_CHANGING_CHARSET(int *_rc) {
-	assert(mybe->server_myds->myconn);
-	MySQL_Data_Stream *myds=mybe->server_myds;
-	MySQL_Connection *myconn=myds->myconn;
-	char msg[128];
-	const MARIADB_CHARSET_INFO *ci = NULL;
-	const char* replace_collation = "";
-	const char* not_supported_collation = "";
-
-	/* Validate that server can support client's charset */
-	if (client_myds->myconn->options.charset >= 255 && myconn->mysql->server_version[0] != '8') {
-		switch(mysql_thread___handle_unknown_charset) {
-			case HANDLE_UNKNOWN_CHARSET__DISCONNECT_CLIENT:
-				snprintf(msg,sizeof(msg),"Can't initialize character set %d",client_myds->myconn->options.charset);
-				proxy_error("Can't initialize character set on %s, %d: Error %d (%s). Closing client connection %s:%d.\n",
-						myconn->parent->address, myconn->parent->port, 2019, msg, client_myds->addr.addr, client_myds->addr.port);
-				myds->destroy_MySQL_Connection_From_Pool(false);
-				myds->fd=0;
-				*_rc=-1;
-				return false;
-			case HANDLE_UNKNOWN_CHARSET__REPLACE_WITH_DEFAULT_VERBOSE:
-				ci = proxysql_find_charset_nr(client_myds->myconn->options.charset);
-				if (ci)	not_supported_collation = ci->name;
-
-				ci = proxysql_find_charset_nr(mysql_thread___default_charset);
-				if (ci)	replace_collation = ci->name;
-
-				proxy_warning("Server doesn't support collation (%d) %s. Replacing it with the configured default (%d) %s. Client %s:%d\n",
-						client_myds->myconn->options.charset, not_supported_collation, 
-						mysql_thread___default_charset, replace_collation, client_myds->addr.addr, client_myds->addr.port);
-
-				client_myds->myconn->options.charset=mysql_thread___default_charset;
-				break;
-			case HANDLE_UNKNOWN_CHARSET__REPLACE_WITH_DEFAULT:
-				client_myds->myconn->options.charset=mysql_thread___default_charset;
-				break;
-			default:
-				proxy_error("Wrong configuration of the handle_unknown_charset\n");
-				break;
-		}
-	}
-
-	myds->DSS=STATE_MARIADB_QUERY;
```

Comment:
I believe this is related to the method `bool MySQL_Session::handler_again___status_CHANGING_CHARSET(int *_rc)`. 

There is no such method any more. New implementation does not execute neither `set charset` nor `set names`. Instead of executing these methods proxysql will keep internally state of the character set variables: `character_set_client`, `character_set_results`, `character_set_connection` and `character_set_database` for each client connection. Proxysql will set directly these variables on backend when multiplexing connection and will not use `set names` as it did before.

For handling `set names` and `character_set_database` this issue will be implemented #2596

Reply:
Ah! Okay, I see, so that part was just the storage of the charset through that commands, thanks for the clarification, I understand the change better now. Should the definition of the method `MySQL_Session::handler_again___status_CHANGING_CHARSET(int *_rc)` be removed then? Thanks you!

Reply:
Fixed